{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "9a75d4d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import pandas as pd\n",
    "\n",
    "from os import walk\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from Bio import AlignIO\n",
    "from CAI import CAI\n",
    "\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "ea6bf050",
   "metadata": {},
   "outputs": [],
   "source": [
    "WORK_PATH = 'aligments_and_sequenses_separated_genes/proteins'\n",
    "MACSE_PATH = '/Users/pitikov_egor/Documents/dora_tree_final/macse_v2.06.jar'\n",
    "PROTEIN_SET = ['atp6', 'co1', 'co2', 'co3', 'cob', 'nad1', 'nad2', 'nad3', 'nad4', 'nad4l', 'nad5', 'nad6']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "e5369cec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_gc_table(ids, path):\n",
    "    t = pd.DataFrame(ids)\n",
    "    t.loc[:, 'gc'] = 5\n",
    "    t.to_csv(path, sep = '\\t', index = False, header = False)\n",
    "    return path\n",
    "\n",
    "def make_ali_prot(fasta_path, fasta_alias):\n",
    "    d_list = []\n",
    "    for rec in SeqIO.parse(f'{fasta_path}/seqs/nt/{fasta_alias}.fasta', 'fasta'):\n",
    "        d_list.append(rec.description)\n",
    "        \n",
    "    gct = make_gc_table(d_list, f'{fasta_path}/seqs/nt/{fasta_alias}_gc.tsv')\n",
    "    \n",
    "    cmd = f'java -jar {MACSE_PATH} -prog alignSequences -gc_file {gct} -out_AA {fasta_path}/ali/aa/{fasta_alias}_ali_aa_codone.fasta -out_NT {fasta_path}/ali/nt/{fasta_alias}_ali_nt_codone.fasta -seq {fasta_path}/seqs/nt/{fasta_alias}.fasta'\n",
    "    p = subprocess.Popen(cmd, shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE)\n",
    "    out, err = p.communicate()\n",
    "            \n",
    "    cmd = f'muscle -in {fasta_path}/seqs/nt/{fasta_alias}.fasta -out {fasta_path}/ali/nt/{fasta_alias}_ali_nt.fasta'\n",
    "    p = subprocess.Popen(cmd, shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE)\n",
    "    out, err = p.communicate()\n",
    "\n",
    "    cmd = f'muscle -in {fasta_path}/seqs/aa/{fasta_alias}.fasta -out {fasta_path}/ali/aa/{fasta_alias}_ali_aa.fasta'\n",
    "    p = subprocess.Popen(cmd, shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE)\n",
    "    out, err = p.communicate()\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "0b4d3530",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6d17bd40672f40799ef95bfadba597f3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Progress in proteins...:   0%|          | 0/12 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for prot_name in tqdm(PROTEIN_SET, desc = 'Progress in proteins...'):\n",
    "    make_ali_prot(WORK_PATH, prot_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "ae65394f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def replase_smbs(file, smb1, smb2):\n",
    "    with open(file, 'r') as read_f:\n",
    "        t = read_f.read()\n",
    "    with open(file, 'w') as file_write:\n",
    "        file_write.write(t.replace(smb1, smb2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "e9737641",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ec9ec71cc8be4b559199fe71f674fe65",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Progress in proteins...:   0%|          | 0/12 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for prot_name in tqdm(PROTEIN_SET, desc = 'Progress in proteins...'):\n",
    "    replase_smbs(f'{WORK_PATH}/ali/aa/{prot_name}_ali_aa_codone.fasta', '!', '-')\n",
    "    replase_smbs(f'{WORK_PATH}/ali/nt/{prot_name}_ali_nt_codone.fasta', '!', '-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "3e08a384",
   "metadata": {},
   "outputs": [],
   "source": [
    "def unite_fasta(fasta_set, fasta_output):\n",
    "    res = {}\n",
    "    tresholds = [0]\n",
    "    for filepath in fasta_set:\n",
    "        orgs = {}\n",
    "        for rec in SeqIO.parse(filepath, 'fasta'):\n",
    "            orgname = ' '.join(rec.description.split(' ')[: -1])\n",
    "            orgs[orgname] = str(rec.seq)\n",
    "        for i in set(orgs.keys()).intersection(set(res.keys())):\n",
    "            res[i] += orgs[i]\n",
    "        for i in orgs.keys() - res.keys():\n",
    "            res[i] = '-' * sum(tresholds) + orgs[i]\n",
    "            print(f'NEW ORGANISM {i} ON {filepath.split(\"/\")[-1]} file')\n",
    "        for i in res.keys() - orgs.keys():\n",
    "            res[i] += '-' * tresholds[-1]\n",
    "            print(f'OLD ORGANISM {i} HAVE NOT BEEN MET IN {filepath.split(\"/\")[-1]} file')\n",
    "            \n",
    "        tresholds.append(len(orgs[orgname]))\n",
    "    return res, tresholds\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "c4f9cee8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--------------------------------------\n",
      "------START NT CODONE ----------------\n",
      "--------------------------------------\n",
      "\n",
      "NEW ORGANISM  MW528029.1 Boccardiella hamata ON atp6_ali_nt_codone.fasta file\n",
      "NEW ORGANISM  MK120303.1 Marenzelleria neglecta ON atp6_ali_nt_codone.fasta file\n",
      "NEW ORGANISM  OQ078742.1 Polydora cf. ciliata DS-2023 ON atp6_ali_nt_codone.fasta file\n",
      "NEW ORGANISM  NC_028711.1 Magelona mirabilis ON atp6_ali_nt_codone.fasta file\n",
      "NEW ORGANISM  NC_061377.1 Polydora hoplura ON atp6_ali_nt_codone.fasta file\n",
      "NEW ORGANISM  MW316633.1 Polydora websteri ON atp6_ali_nt_codone.fasta file\n",
      "NEW ORGANISM  OK032597.1 Lindaspio sp. i ZZ-2021 ON atp6_ali_nt_codone.fasta file\n",
      "NEW ORGANISM  MW316635.1 Polydora brevipalpa ON atp6_ali_nt_codone.fasta file\n",
      "\n",
      "--------------------------------------\n",
      "------START AA CODONE ----------------\n",
      "--------------------------------------\n",
      "\n",
      "NEW ORGANISM  MW528029.1 Boccardiella hamata ON atp6_ali_aa_codone.fasta file\n",
      "NEW ORGANISM  MK120303.1 Marenzelleria neglecta ON atp6_ali_aa_codone.fasta file\n",
      "NEW ORGANISM  OQ078742.1 Polydora cf. ciliata DS-2023 ON atp6_ali_aa_codone.fasta file\n",
      "NEW ORGANISM  NC_028711.1 Magelona mirabilis ON atp6_ali_aa_codone.fasta file\n",
      "NEW ORGANISM  NC_061377.1 Polydora hoplura ON atp6_ali_aa_codone.fasta file\n",
      "NEW ORGANISM  MW316633.1 Polydora websteri ON atp6_ali_aa_codone.fasta file\n",
      "NEW ORGANISM  OK032597.1 Lindaspio sp. i ZZ-2021 ON atp6_ali_aa_codone.fasta file\n",
      "NEW ORGANISM  MW316635.1 Polydora brevipalpa ON atp6_ali_aa_codone.fasta file\n",
      "\n",
      "--------------------------------------\n",
      "------START NT -----------------------\n",
      "--------------------------------------\n",
      "\n",
      "NEW ORGANISM  MW528029.1 Boccardiella hamata ON atp6_ali_nt.fasta file\n",
      "NEW ORGANISM  MK120303.1 Marenzelleria neglecta ON atp6_ali_nt.fasta file\n",
      "NEW ORGANISM  OQ078742.1 Polydora cf. ciliata DS-2023 ON atp6_ali_nt.fasta file\n",
      "NEW ORGANISM  NC_028711.1 Magelona mirabilis ON atp6_ali_nt.fasta file\n",
      "NEW ORGANISM  OK032597.1 Lindaspio sp. i ZZ-2021 ON atp6_ali_nt.fasta file\n",
      "NEW ORGANISM  NC_061377.1 Polydora hoplura ON atp6_ali_nt.fasta file\n",
      "NEW ORGANISM  MW316633.1 Polydora websteri ON atp6_ali_nt.fasta file\n",
      "NEW ORGANISM  MW316635.1 Polydora brevipalpa ON atp6_ali_nt.fasta file\n",
      "\n",
      "--------------------------------------\n",
      "------START AA -----------------------\n",
      "--------------------------------------\n",
      "\n",
      "NEW ORGANISM  MW528029.1 Boccardiella hamata ON atp6_ali_aa.fasta file\n",
      "NEW ORGANISM  MK120303.1 Marenzelleria neglecta ON atp6_ali_aa.fasta file\n",
      "NEW ORGANISM  OQ078742.1 Polydora cf. ciliata DS-2023 ON atp6_ali_aa.fasta file\n",
      "NEW ORGANISM  NC_028711.1 Magelona mirabilis ON atp6_ali_aa.fasta file\n",
      "NEW ORGANISM  OK032597.1 Lindaspio sp. i ZZ-2021 ON atp6_ali_aa.fasta file\n",
      "NEW ORGANISM  NC_061377.1 Polydora hoplura ON atp6_ali_aa.fasta file\n",
      "NEW ORGANISM  MW316633.1 Polydora websteri ON atp6_ali_aa.fasta file\n",
      "NEW ORGANISM  MW316635.1 Polydora brevipalpa ON atp6_ali_aa.fasta file\n"
     ]
    }
   ],
   "source": [
    "for prot_name in PROTEIN_SET:\n",
    "    replase_smbs(f'{WORK_PATH}/ali/aa/{prot_name}_ali_aa_codone.fasta', '!', '-')\n",
    "    replase_smbs(f'{WORK_PATH}/ali/nt/{prot_name}_ali_nt_codone.fasta', '!', '-')\n",
    "    \n",
    "nt_set_codone = [f'{WORK_PATH}/ali/nt/{prot_name}_ali_nt_codone.fasta' for prot_name in PROTEIN_SET]\n",
    "aa_set_codone = [f'{WORK_PATH}/ali/aa/{prot_name}_ali_aa_codone.fasta' for prot_name in PROTEIN_SET]\n",
    "\n",
    "nt_set = [f'{WORK_PATH}/ali/nt/{prot_name}_ali_nt.fasta' for prot_name in PROTEIN_SET]\n",
    "aa_set = [f'{WORK_PATH}/ali/aa/{prot_name}_ali_aa.fasta' for prot_name in PROTEIN_SET]\n",
    "\n",
    "print('''\n",
    "--------------------------------------\n",
    "------START NT CODONE ----------------\n",
    "--------------------------------------\n",
    "''')\n",
    "  = unite_fasta(nt_set_codone, f'{WORK_PATH}/ali/nt/total_ali_nt_codone.fasta')\n",
    "print('''\n",
    "--------------------------------------\n",
    "------START AA CODONE ----------------\n",
    "--------------------------------------\n",
    "''')\n",
    "aa_codone, aa_c_tr = unite_fasta(aa_set_codone, f'{WORK_PATH}/ali/aa/total_ali_aa_codone.fasta')\n",
    "print('''\n",
    "--------------------------------------\n",
    "------START NT -----------------------\n",
    "--------------------------------------\n",
    "''')\n",
    "nt, nt_tr = unite_fasta(nt_set, f'{WORK_PATH}/ali/nt/total_ali_nt.fasta')\n",
    "print('''\n",
    "--------------------------------------\n",
    "------START AA -----------------------\n",
    "--------------------------------------\n",
    "''')\n",
    "aa, aa_tr = unite_fasta(aa_set, f'{WORK_PATH}/ali/aa/total_ali_aa.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "4e24d442",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_fasta_and_nexus(ali_dict, \n",
    "                         tresholds, \n",
    "                         outdir, \n",
    "                         alias, \n",
    "                         feat_type = 'DNA', \n",
    "                         model = 'GTR+I+G'):\n",
    "    to_file = [SeqRecord(id = '', description = i, seq = Seq(ali_dict[i])) for i in ali_dict.keys()] \n",
    "    with open(f'{outdir}/{alias}.fasta', 'w') as file:\n",
    "        SeqIO.write(to_file, file, 'fasta')\n",
    "    feat_num = len(tresholds)\n",
    "    with open(f'{outdir}/{alias}.nex', 'w') as file:\n",
    "        file.write('#nexus\\n')\n",
    "        file.write('begin sets;\\n')\n",
    "        last_en = 0\n",
    "        for pos, val in enumerate(tresholds, start = 1):\n",
    "            if pos<feat_num:\n",
    "                st = 1 + last_en\n",
    "                en = tresholds[pos] + st - 1\n",
    "                file.write(f'\\tcharset part{pos} = {st}-{en};\\n')\n",
    "                last_en = en\n",
    "        file.write(f'\\tcharpartition mine =')\n",
    "        for i in range(1, pos-1):\n",
    "            file.write(f' {model}:part{i},')\n",
    "        file.write(f' {model}:part{pos-1};\\n')\n",
    "        file.write('end;')\n",
    "    return f'{outdir}/{alias}.fasta', f'{outdir}/{alias}.nex'\n",
    "\n",
    "def make_tree(ali_dict, \n",
    "              tresholds, \n",
    "              outdir, \n",
    "              alias, \n",
    "              feat_type = 'DNA', \n",
    "              model = 'GTR+I+G', \n",
    "              finder = 'MFP+MERGE', \n",
    "              b = None, \n",
    "              mode = 'p'):\n",
    "    fasta, nexus = make_fasta_and_nexus(ali_dict, \n",
    "                         tresholds, \n",
    "                         outdir, \n",
    "                         alias, \n",
    "                         feat_type = feat_type,\n",
    "                         model = model)\n",
    "    if b:\n",
    "        cmd = f'iqtree -s {fasta} -st {feat_type} -m {finder} -B {b} -{mode} {nexus} --mlrate'\n",
    "    else:\n",
    "        cmd = f'iqtree -s {fasta} -st {feat_type} -m {finder} -{mode} {nexus} --mlrate'\n",
    "    p = subprocess.Popen(cmd, shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE)\n",
    "    out, err = p.communicate()\n",
    "    return out, err"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "e90d924c",
   "metadata": {},
   "outputs": [],
   "source": [
    "models = ['HKY', 'JC', 'F81', 'K2P', 'K3P', 'K81uf', 'TNef', 'TIM', 'TIMef', 'TVM', ', TVMef', 'SYM', 'GTR'] \n",
    "models = models + [f'{i}+I+G' for i in models]\n",
    "cfeats = ['TESTONLY', 'TEST', 'MF', 'MFP']\n",
    "tfeats = ['LM', 'LMRY', 'LMWS', 'LMMK', 'LMSS', '']\n",
    "feats = [f'{i}+{j}' for i in cfeats for j in tfeats]\n",
    "modes = ['p', 'q']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25f3ac0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "47f914097de3418e9cc769e1517a8b61",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import itertools as it\n",
    "cmd = f'mkdir -p test_trees_params'\n",
    "p = subprocess.Popen(cmd, shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE)\n",
    "out, err = p.communicate()\n",
    "for model_type, finder_type, mode in tqdm(it.product(models, feats, modes)):\n",
    "    cmd = f'mkdir -p test_trees_params/{model_type}_{finder_type}_{mode}'\n",
    "    p = subprocess.Popen(cmd, shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE)\n",
    "    out, err = p.communicate()\n",
    "    out, err = make_tree(nt_codone, \n",
    "              nt_c_tr, \n",
    "              f'test_trees_params/{model_type}_{finder_type}_{mode}', \n",
    "              alias = 'codone_all', \n",
    "              model = model_type, \n",
    "              finder = finder_type, \n",
    "              b = 1500)\n",
    "    out = out.decode()\n",
    "    err = err.decode()\n",
    "    with open(f'test_trees_params/{model_type}_{finder_type}_{mode}/out_prog.txt', 'w') as out_file:\n",
    "        out_file.write(out)\n",
    "    with open(f'test_trees_params/{model_type}_{finder_type}_{mode}/err_prog.txt', 'w') as err_file:\n",
    "        err_file.write(err)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "918950d0",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 2.2.0.3 COVID-edition for Mac OS X 64-bit built Sep  5 2022\r\n",
      "Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,\r\n",
      "Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams, Ly Trong Nhan.\r\n",
      "\r\n",
      "Usage: iqtree [-s ALIGNMENT] [-p PARTITION] [-m MODEL] [-t TREE] ...\r\n",
      "\r\n",
      "GENERAL OPTIONS:\r\n",
      "  -h, --help           Print (more) help usages\r\n",
      "  -s FILE[,...,FILE]   PHYLIP/FASTA/NEXUS/CLUSTAL/MSF alignment file(s)\r\n",
      "  -s DIR               Directory of alignment files\r\n",
      "  --seqtype STRING     BIN, DNA, AA, NT2AA, CODON, MORPH (default: auto-detect)\r\n",
      "  -t FILE|PARS|RAND    Starting tree (default: 99 parsimony and BIONJ)\r\n",
      "  -o TAX[,...,TAX]     Outgroup taxon (list) for writing .treefile\r\n",
      "  --prefix STRING      Prefix for all output files (default: aln/partition)\r\n",
      "  --seed NUM           Random seed number, normally used for debugging purpose\r\n",
      "  --safe               Safe likelihood kernel to avoid numerical underflow\r\n",
      "  --mem NUM[G|M|%]     Maximal RAM usage in GB | MB | %\r\n",
      "  --runs NUM           Number of indepedent runs (default: 1)\r\n",
      "  -v, --verbose        Verbose mode, printing more messages to screen\r\n",
      "  -V, --version        Display version number\r\n",
      "  --quiet              Quiet mode, suppress printing to screen (stdout)\r\n",
      "  -fconst f1,...,fN    Add constant patterns into alignment (N=no. states)\r\n",
      "  --epsilon NUM        Likelihood epsilon for parameter estimate (default 0.01)\r\n",
      "  -T NUM|AUTO          No. cores/threads or AUTO-detect (default: 1)\r\n",
      "  --threads-max NUM    Max number of threads for -T AUTO (default: all cores)\r\n",
      "  --export-alisim-cmd  Export a command-line from the inferred tree and model params\r\n",
      "                       to simulate new MSAs with AliSim\r\n",
      "\r\n",
      "CHECKPOINT:\r\n",
      "  --redo               Redo both ModelFinder and tree search\r\n",
      "  --redo-tree          Restore ModelFinder and only redo tree search\r\n",
      "  --undo               Revoke finished run, used when changing some options\r\n",
      "  --cptime NUM         Minimum checkpoint interval (default: 60 sec and adapt)\r\n",
      "\r\n",
      "PARTITION MODEL:\r\n",
      "  -p FILE|DIR          NEXUS/RAxML partition file or directory with alignments\r\n",
      "                       Edge-linked proportional partition model\r\n",
      "  -q FILE|DIR          Like -p but edge-linked equal partition model \r\n",
      "  -Q FILE|DIR          Like -p but edge-unlinked partition model\r\n",
      "  -S FILE|DIR          Like -p but separate tree inference\r\n",
      "  --subsample NUM      Randomly sub-sample partitions (negative for complement)\r\n",
      "  --subsample-seed NUM Random number seed for --subsample\r\n",
      "\r\n",
      "LIKELIHOOD/QUARTET MAPPING:\r\n",
      "  --lmap NUM           Number of quartets for likelihood mapping analysis\r\n",
      "  --lmclust FILE       NEXUS file containing clusters for likelihood mapping\r\n",
      "  --quartetlh          Print quartet log-likelihoods to .quartetlh file\r\n",
      "\r\n",
      "TREE SEARCH ALGORITHM:\r\n",
      "  --ninit NUM          Number of initial parsimony trees (default: 100)\r\n",
      "  --ntop NUM           Number of top initial trees (default: 20)\r\n",
      "  --nbest NUM          Number of best trees retained during search (defaut: 5)\r\n",
      "  -n NUM               Fix number of iterations to stop (default: OFF)\r\n",
      "  --nstop NUM          Number of unsuccessful iterations to stop (default: 100)\r\n",
      "  --perturb NUM        Perturbation strength for randomized NNI (default: 0.5)\r\n",
      "  --radius NUM         Radius for parsimony SPR search (default: 6)\r\n",
      "  --allnni             Perform more thorough NNI search (default: OFF)\r\n",
      "  -g FILE              (Multifurcating) topological constraint tree file\r\n",
      "  --fast               Fast search to resemble FastTree\r\n",
      "  --polytomy           Collapse near-zero branches into polytomy\r\n",
      "  --tree-fix           Fix -t tree (no tree search performed)\r\n",
      "  --treels             Write locally optimal trees into .treels file\r\n",
      "  --show-lh            Compute tree likelihood without optimisation\r\n",
      "  --terrace            Check if the tree lies on a phylogenetic terrace\r\n",
      "\r\n",
      "ULTRAFAST BOOTSTRAP/JACKKNIFE:\r\n",
      "  -B, --ufboot NUM     Replicates for ultrafast bootstrap (>=1000)\r\n",
      "  -J, --ufjack NUM     Replicates for ultrafast jackknife (>=1000)\r\n",
      "  --jack-prop NUM      Subsampling proportion for jackknife (default: 0.5)\r\n",
      "  --sampling STRING    GENE|GENESITE resampling for partitions (default: SITE)\r\n",
      "  --boot-trees         Write bootstrap trees to .ufboot file (default: none)\r\n",
      "  --wbtl               Like --boot-trees but also writing branch lengths\r\n",
      "  --nmax NUM           Maximum number of iterations (default: 1000)\r\n",
      "  --nstep NUM          Iterations for UFBoot stopping rule (default: 100)\r\n",
      "  --bcor NUM           Minimum correlation coefficient (default: 0.99)\r\n",
      "  --beps NUM           RELL epsilon to break tie (default: 0.5)\r\n",
      "  --bnni               Optimize UFBoot trees by NNI on bootstrap alignment\r\n",
      "\r\n",
      "NON-PARAMETRIC BOOTSTRAP/JACKKNIFE:\r\n",
      "  -b, --boot NUM       Replicates for bootstrap + ML tree + consensus tree\r\n",
      "  -j, --jack NUM       Replicates for jackknife + ML tree + consensus tree\r\n",
      "  --jack-prop NUM      Subsampling proportion for jackknife (default: 0.5)\r\n",
      "  --bcon NUM           Replicates for bootstrap + consensus tree\r\n",
      "  --bonly NUM          Replicates for bootstrap only\r\n",
      "  --tbe                Transfer bootstrap expectation\r\n",
      "\r\n",
      "SINGLE BRANCH TEST:\r\n",
      "  --alrt NUM           Replicates for SH approximate likelihood ratio test\r\n",
      "  --alrt 0             Parametric aLRT test (Anisimova and Gascuel 2006)\r\n",
      "  --abayes             approximate Bayes test (Anisimova et al. 2011)\r\n",
      "  --lbp NUM            Replicates for fast local bootstrap probabilities\r\n",
      "\r\n",
      "MODEL-FINDER:\r\n",
      "  -m TESTONLY          Standard model selection (like jModelTest, ProtTest)\r\n",
      "  -m TEST              Standard model selection followed by tree inference\r\n",
      "  -m MF                Extended model selection with FreeRate heterogeneity\r\n",
      "  -m MFP               Extended model selection followed by tree inference\r\n",
      "  -m ...+LM            Additionally test Lie Markov models\r\n",
      "  -m ...+LMRY          Additionally test Lie Markov models with RY symmetry\r\n",
      "  -m ...+LMWS          Additionally test Lie Markov models with WS symmetry\r\n",
      "  -m ...+LMMK          Additionally test Lie Markov models with MK symmetry\r\n",
      "  -m ...+LMSS          Additionally test strand-symmetric models\r\n",
      "  --mset STRING        Restrict search to models supported by other programs\r\n",
      "                       (raxml, phyml, mrbayes, beast1 or beast2)\r\n",
      "  --mset STR,...       Comma-separated model list (e.g. -mset WAG,LG,JTT)\r\n",
      "  --msub STRING        Amino-acid model source\r\n",
      "                       (nuclear, mitochondrial, chloroplast or viral)\r\n",
      "  --mfreq STR,...      List of state frequencies\r\n",
      "  --mrate STR,...      List of rate heterogeneity among sites\r\n",
      "                       (e.g. -mrate E,I,G,I+G,R is used for -m MF)\r\n",
      "  --cmin NUM           Min categories for FreeRate model [+R] (default: 2)\r\n",
      "  --cmax NUM           Max categories for FreeRate model [+R] (default: 10)\r\n",
      "  --merit AIC|AICc|BIC  Akaike|Bayesian information criterion (default: BIC)\r\n",
      "  --mtree              Perform full tree search for every model\r\n",
      "  --madd STR,...       List of mixture models to consider\r\n",
      "  --mdef FILE          Model definition NEXUS file (see Manual)\r\n",
      "  --modelomatic        Find best codon/protein/DNA models (Whelan et al. 2015)\r\n",
      "\r\n",
      "PARTITION-FINDER:\r\n",
      "  --merge              Merge partitions to increase model fit\r\n",
      "  --merge greedy|rcluster|rclusterf\r\n",
      "                       Set merging algorithm (default: rclusterf)\r\n",
      "  --merge-model 1|all  Use only 1 or all models for merging (default: 1)\r\n",
      "  --merge-model STR,...\r\n",
      "                       Comma-separated model list for merging\r\n",
      "  --merge-rate 1|all   Use only 1 or all rate heterogeneity (default: 1)\r\n",
      "  --merge-rate STR,...\r\n",
      "                       Comma-separated rate list for merging\r\n",
      "  --rcluster NUM       Percentage of partition pairs for rcluster algorithm\r\n",
      "  --rclusterf NUM      Percentage of partition pairs for rclusterf algorithm\r\n",
      "  --rcluster-max NUM   Max number of partition pairs (default: 10*partitions)\r\n",
      "\r\n",
      "SUBSTITUTION MODEL:\r\n",
      "  -m STRING            Model name string (e.g. GTR+F+I+G)\r\n",
      "                 DNA:  HKY (default), JC, F81, K2P, K3P, K81uf, TN/TrN, TNef,\r\n",
      "                       TIM, TIMef, TVM, TVMef, SYM, GTR, or 6-digit model\r\n",
      "                       specification (e.g., 010010 = HKY)\r\n",
      "             Protein:  LG (default), Poisson, cpREV, mtREV, Dayhoff, mtMAM,\r\n",
      "                       JTT, WAG, mtART, mtZOA, VT, rtREV, DCMut, PMB, HIVb,\r\n",
      "                       HIVw, JTTDCMut, FLU, Blosum62, GTR20, mtMet, mtVer, mtInv, FLAVI,\r\n",
      "\t\t\tQ.LG, Q.pfam, Q.pfam_gb, Q.bird, Q.mammal, Q.insect, Q.plant, Q.yeast\r\n",
      "     Protein mixture:  C10,...,C60, EX2, EX3, EHO, UL2, UL3, EX_EHO, LG4M, LG4X\r\n",
      "              Binary:  JC2 (default), GTR2\r\n",
      "     Empirical codon:  KOSI07, SCHN05\r\n",
      "   Mechanistic codon:  GY (default), MG, MGK, GY0K, GY1KTS, GY1KTV, GY2K,\r\n",
      "                       MG1KTS, MG1KTV, MG2K\r\n",
      "Semi-empirical codon:  XX_YY where XX is empirical and YY is mechanistic model\r\n",
      "      Morphology/SNP:  MK (default), ORDERED, GTR\r\n",
      "      Lie Markov DNA:  1.1, 2.2b, 3.3a, 3.3b, 3.3c, 3.4, 4.4a, 4.4b, 4.5a,\r\n",
      "                       4.5b, 5.6a, 5.6b, 5.7a, 5.7b, 5.7c, 5.11a, 5.11b, 5.11c,\r\n",
      "                       5.16, 6.6, 6.7a, 6.7b, 6.8a, 6.8b, 6.17a, 6.17b, 8.8,\r\n",
      "                       8.10a, 8.10b, 8.16, 8.17, 8.18, 9.20a, 9.20b, 10.12,\r\n",
      "                       10.34, 12.12 (optionally prefixed by RY, WS or MK)\r\n",
      "      Non-reversible:  STRSYM (strand symmetric model, equiv. WS6.6),\r\n",
      "                       NONREV, UNREST (unrestricted model, equiv. 12.12)\r\n",
      "           Otherwise:  Name of file containing user-model parameters\r\n",
      "\r\n",
      "STATE FREQUENCY:\r\n",
      "  -m ...+F             Empirically counted frequencies from alignment\r\n",
      "  -m ...+FO            Optimized frequencies by maximum-likelihood\r\n",
      "  -m ...+FQ            Equal frequencies\r\n",
      "  -m ...+FRY           For DNA, freq(A+G)=1/2=freq(C+T)\r\n",
      "  -m ...+FWS           For DNA, freq(A+T)=1/2=freq(C+G)\r\n",
      "  -m ...+FMK           For DNA, freq(A+C)=1/2=freq(G+T)\r\n",
      "  -m ...+Fabcd         4-digit constraint on ACGT frequency\r\n",
      "                       (e.g. +F1221 means f_A=f_T, f_C=f_G)\r\n",
      "  -m ...+FU            Amino-acid frequencies given protein matrix\r\n",
      "  -m ...+F1x4          Equal NT frequencies over three codon positions\r\n",
      "  -m ...+F3x4          Unequal NT frequencies over three codon positions\r\n",
      "\r\n",
      "RATE HETEROGENEITY AMONG SITES:\r\n",
      "  -m ...+I             A proportion of invariable sites\r\n",
      "  -m ...+G[n]          Discrete Gamma model with n categories (default n=4)\r\n",
      "  -m ...*G[n]          Discrete Gamma model with unlinked model parameters\r\n",
      "  -m ...+I+G[n]        Invariable sites plus Gamma model with n categories\r\n",
      "  -m ...+R[n]          FreeRate model with n categories (default n=4)\r\n",
      "  -m ...*R[n]          FreeRate model with unlinked model parameters\r\n",
      "  -m ...+I+R[n]        Invariable sites plus FreeRate model with n categories\r\n",
      "  -m ...+Hn            Heterotachy model with n classes\r\n",
      "  -m ...*Hn            Heterotachy model with n classes and unlinked parameters\r\n",
      "  --alpha-min NUM      Min Gamma shape parameter for site rates (default: 0.02)\r\n",
      "  --gamma-median       Median approximation for +G site rates (default: mean)\r\n",
      "  --rate               Write empirical Bayesian site rates to .rate file\r\n",
      "  --mlrate             Write maximum likelihood site rates to .mlrate file\r\n",
      "\r\n",
      "POLYMORPHISM AWARE MODELS (PoMo):\r\n",
      "  -s FILE              Input counts file (see manual)\r\n",
      "  -m ...+P             DNA substitution model (see above) used with PoMo\r\n",
      "  -m ...+N<POPSIZE>    Virtual population size (default: 9)\r\n",
      "  -m ...+WB|WH|S]      Weighted binomial sampling\r\n",
      "  -m ...+WH            Weighted hypergeometric sampling\r\n",
      "  -m ...+S             Sampled sampling\r\n",
      "  -m ...+G[n]          Discrete Gamma rate with n categories (default n=4)\r\n",
      "\r\n",
      "COMPLEX MODELS:\r\n",
      "  -m \"MIX{m1,...,mK}\"  Mixture model with K components\r\n",
      "  -m \"FMIX{f1,...fK}\"  Frequency mixture model with K components\r\n",
      "  --mix-opt            Optimize mixture weights (default: detect)\r\n",
      "  -m ...+ASC           Ascertainment bias correction\r\n",
      "  --tree-freq FILE     Input tree to infer site frequency model\r\n",
      "  --site-freq FILE     Input site frequency model file\r\n",
      "  --freq-max           Posterior maximum instead of mean approximation\r\n",
      "\r\n",
      "TREE TOPOLOGY TEST:\r\n",
      "  --trees FILE         Set of trees to evaluate log-likelihoods\r\n",
      "  --test NUM           Replicates for topology test\r\n",
      "  --test-weight        Perform weighted KH and SH tests\r\n",
      "  --test-au            Approximately unbiased (AU) test (Shimodaira 2002)\r\n",
      "  --sitelh             Write site log-likelihoods to .sitelh file\r\n",
      "\r\n",
      "ANCESTRAL STATE RECONSTRUCTION:\r\n",
      "  --ancestral          Ancestral state reconstruction by empirical Bayes\r\n",
      "  --asr-min NUM        Min probability of ancestral state (default: equil freq)\r\n",
      "\r\n",
      "TEST OF SYMMETRY:\r\n",
      "  --symtest               Perform three tests of symmetry\r\n",
      "  --symtest-only          Do --symtest then exist\r\n",
      "  --symtest-remove-bad    Do --symtest and remove bad partitions\r\n",
      "  --symtest-remove-good   Do --symtest and remove good partitions\r\n",
      "  --symtest-type MAR|INT  Use MARginal/INTernal test when removing partitions\r\n",
      "  --symtest-pval NUMER    P-value cutoff (default: 0.05)\r\n",
      "  --symtest-keep-zero     Keep NAs in the tests\r\n",
      "\r\n",
      "CONCORDANCE FACTOR ANALYSIS:\r\n",
      "  -t FILE              Reference tree to assign concordance factor\r\n",
      "  --gcf FILE           Set of source trees for gene concordance factor (gCF)\r\n",
      "  --df-tree            Write discordant trees associated with gDF1\r\n",
      "  --scf NUM            Number of quartets for site concordance factor (sCF)\r\n",
      "  -s FILE              Sequence alignment for --scf\r\n",
      "  -p FILE|DIR          Partition file or directory for --scf\r\n",
      "  --cf-verbose         Write CF per tree/locus to cf.stat_tree/_loci\r\n",
      "  --cf-quartet         Write sCF for all resampled quartets to .cf.quartet\r\n",
      "\r\n",
      "ALISIM: ALIGNMENT SIMULATOR\r\n",
      "\r\n",
      "Usage: iqtree --alisim <OUTPUT_PREFIX> [-m MODEL] [-t TREE] ...\r\n",
      "\r\n",
      "  --alisim OUTPUT_ALIGNMENT Activate AliSim and specify the output alignment filename\r\n",
      "  -t TREE_FILE              Set the input tree file name\r\n",
      "  --length LENGTH           Set the length of the root sequence\r\n",
      "  --num-alignments NUMBER   Set the number of output datasets\r\n",
      "  --seqtype STRING          BIN, DNA, AA, CODON, MORPH{NUM_STATES} (default: auto-detect)\r\n",
      "                            For morphological data, 0<NUM_STATES<=32\r\n",
      "  --m MODEL_STRING          Specify the evolutionary model. See Manual for more detail\r\n",
      "  --mdef FILE               Name of a NEXUS model file to define new models (see Manual)\r\n",
      "  --fundi TAXA_LIST,RHO     Specify a list of taxa, and Rho (Fundi weight) for FunDi model\r\n",
      "  --indel <INS>,<DEL>       Set the insertion and deletion rate of the indel model,\r\n",
      "                            relative to the substitution rate\r\n",
      "  --indel-size <INS_DIS>,<DEL_DIS> Set the insertion and deletion size distributions\r\n",
      "  --sub-level-mixture       Enable the feature to simulate substitution-level mixture model\r\n",
      "  --no-unaligned            Disable outputing a file of unaligned sequences \r\n",
      "                            when using indel models\r\n",
      "  --root-seq FILE,SEQ_NAME  Specify the root sequence from an alignment\r\n",
      "  -s FILE                   Specify the input sequence alignment\r\n",
      "  --no-copy-gaps            Disable copying gaps from input alignment (default: false)\r\n",
      "  --site-freq <OPTION>      Specify the option (MEAN (default), or SAMPLING, or MODEL)\r\n",
      "                            to mimic the site-frequencies for mixture models from\r\n",
      "                            the input alignment (see Manual)\r\n",
      "  --site-rate <OPTION>      Specify the option (MEAN (default), or SAMPLING, or MODEL)\r\n",
      "                            to mimic the discrete rate heterogeneity from\r\n",
      "                            the input alignment (see Manual)\r\n",
      "  -t RANDOM{MODEL,NUM_TAXA} Specify the model and the number of taxa to generate a random tree\r\n",
      "  -rlen MIN MEAN MAX        Specify three numbers: minimum, mean and maximum branch lengths\r\n",
      "                            when generating a random tree\r\n",
      "  -p FILE                   NEXUS/RAxML partition file\r\n",
      "                            Edge-linked proportional partition model\r\n",
      "  -q FILE                   Like -p but edge-linked equal partition model \r\n",
      "  -Q FILE                   Like -p but edge-unlinked partition model\r\n",
      "  --distribution FILE       Supply a definition file of distributions,\r\n",
      "                            which could be used to generate random model parameters\r\n",
      "  --branch-distribution DIS Specify a distribution, from which branch lengths of the input trees\r\n",
      "                            are randomly generated and overridden.\r\n",
      "  --branch-scale SCALE      Specify a value to scale all branch lengths\r\n",
      "  --single-output           Output all alignments into a single file\r\n",
      "  --write-all               Enable outputting internal sequences\r\n",
      "  --seed NUM                Random seed number (default: CPU clock)\r\n",
      "                            Be careful to make the AliSim reproducible,\r\n",
      "                            users should specify the seed number\r\n",
      "  -gz                       Enable output compression but taking longer running time\r\n",
      "  -af phy|fasta             Set the output format (default: phylip)\r\n",
      "  User Manual is available at http://www.iqtree.org/doc/alisim\r\n",
      "\r\n",
      "ANALYSIS WITH GENTRIUS ALGORITHM:\r\n",
      "  --gentrius FILE      File must contain either a single species-tree or a set of subtrees.\r\n",
      "  -pr_ab_matrix FILE   Presence-absence matrix of loci coverage.\r\n",
      "  -s FILE              PHYLIP/FASTA/NEXUS/CLUSTAL/MSF alignment file(s)\r\n",
      "  -p FILE              NEXUS/RAxML partition file\r\n",
      "  -g_stop_t NUM        Stop after NUM species-trees were generated, or use 0 to turn off this stopping rule. Default: 1MLN trees.\r\n",
      "  -g_stop_i NUM        Stop after NUM intermediate trees were visited, or use 0 to turn off this stopping rule. Default: 10MLN trees.\r\n",
      "  -g_stop_h NUM        Stop after NUM hours (CPU time), or use 0 to turn off this stopping rule. Default: 7 days.\r\n",
      "  -g_non_stop          Turn off all stopping rules.\r\n",
      "  -g_query FILE        Species-trees to test for identical set of subtrees.\r\n",
      "  -g_print             Write all generated species-trees. WARNING: there might be millions of trees!\r\n",
      "  -g_print_lim NUM     Limit on the number of species-trees to be written.\r\n",
      "  -g_print_induced     Write induced partition subtrees.\r\n",
      "  -g_print_m           Write presence-absence matrix.\r\n",
      "  -g_rm_leaves NUM     Invoke reverse analysis for complex datasets.\r\n",
      "\r\n",
      "TIME TREE RECONSTRUCTION:\r\n",
      "  --date FILE          File containing dates of tips or ancestral nodes\r\n",
      "  --date TAXNAME       Extract dates from taxon names after last '|'\r\n",
      "  --date-tip STRING    Tip dates as a real number or YYYY-MM-DD\r\n",
      "  --date-root STRING   Root date as a real number or YYYY-MM-DD\r\n",
      "  --date-ci NUM        Number of replicates to compute confidence interval\r\n",
      "  --clock-sd NUM       Std-dev for lognormal relaxed clock (default: 0.2)\r\n",
      "  --date-no-outgroup   Exclude outgroup from time tree\r\n",
      "  --date-outlier NUM   Z-score cutoff to remove outlier tips/nodes (e.g. 3)\r\n",
      "  --date-options \"..\"  Extra options passing directly to LSD2\r\n",
      "  --dating STRING      Dating method: LSD for least square dating (default)\r\n",
      "\r\n",
      "\r\n",
      "CONSENSUS RECONSTRUCTION:\r\n",
      "  -t FILE              Set of input trees for consensus reconstruction\r\n",
      "  --sup-min NUM        Min split support, 0.5 for majority-rule consensus\r\n",
      "                       (default: 0, extended consensus)\r\n",
      "  --burnin NUM         Burnin number of trees to ignore\r\n",
      "  --con-tree           Compute consensus tree to .contree file\r\n",
      "  --con-net            Computing consensus network to .nex file\r\n",
      "  --support FILE       Assign support values into this tree from -t trees\r\n",
      "  --suptag STRING      Node name (or ALL) to assign tree IDs where node occurs\r\n",
      "\r\n",
      "TREE DISTANCE BY ROBINSON-FOULDS (RF) METRIC:\r\n",
      "  --tree-dist-all      Compute all-to-all RF distances for -t trees\r\n",
      "  --tree-dist FILE     Compute RF distances between -t trees and this set\r\n",
      "  --tree-dist2 FILE    Like -rf but trees can have unequal taxon sets\r\n",
      "\r\n",
      "GENERATING RANDOM TREES:\r\n",
      "  -r NUM               No. taxa for Yule-Harding random tree\r\n",
      "  --rand UNI|CAT|BAL   UNIform | CATerpillar | BALanced random tree\r\n",
      "  --rlen NUM NUM NUM   min, mean, and max random branch lengths\r\n",
      "\r\n",
      "MISCELLANEOUS:\r\n",
      "  --keep-ident         Keep identical sequences (default: remove & finally add)\r\n",
      "  -blfix               Fix branch lengths of user tree passed via -te\r\n",
      "  -blscale             Scale branch lengths of user tree passed via -t\r\n",
      "  -blmin               Min branch length for optimization (default 0.000001)\r\n",
      "  -blmax               Max branch length for optimization (default 100)\r\n",
      "  -wslr                Write site log-likelihoods per rate category\r\n",
      "  -wslm                Write site log-likelihoods per mixture class\r\n",
      "  -wslmr               Write site log-likelihoods per mixture+rate class\r\n",
      "  -wspr                Write site probabilities per rate category\r\n",
      "  -wspm                Write site probabilities per mixture class\r\n",
      "  -wspmr               Write site probabilities per mixture+rate class\r\n",
      "  --partlh             Write partition log-likelihoods to .partlh file\r\n",
      "  --no-outfiles        Suppress printing output files\r\n",
      "  --eigenlib           Use Eigen3 library\r\n",
      "  -alninfo             Print alignment sites statistics to .alninfo\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!iqtree --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "4c77f5e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in zip(nt_codone.values()):\n",
    "    if set(i) == set('-'): print(i)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
