# Multiple sequence alignments used to build phylogenetic tress
- codon_all.fasta – concatenated codon alignment of 12 standard protein-coding genes from Spionida and Magelona mirabilis mitogenomes used to construct a phylogenetic tree for Spionida mitogenomes. ATP8 was excluded from the analysis. The alignments were performed using an MSA algorithm integrated into the MACSE tool version 0.9b1 and then concatenated.
- 16S_ali_totree_final.fasta – multiple sequence alignment of 16S sequences of Polydora cf. ciliata and related Annelida species which was used to construct a relevant phylogenetic tree. Nucleotide alignment was performed using the Muscle algorithm with default parameters (Edgar 2004) integrated in MEGA11 software (Tamura et al. 2021). 
- 18S_ali_totree_final.fasta –  multiple sequence alignment of 18S sequences of Polydora cf. ciliata and related Annelida species which was used to construct a relevant phylogenetic tree. Nucleotide alignment was performed using the Muscle algorithm with default parameters (Edgar 2004) integrated in MEGA11 software (Tamura et al. 2021).
- 28S_ali_totree_final.fasta – multiple sequence alignment of 28S sequences of Polydora cf. ciliata and related Annelida species which was used to construct a relevant phylogenetic tree. Nucleotide alignment was performed using the Muscle algorithm with default parameters (Edgar 2004) integrated in MEGA11 software (Tamura et al. 2021). 
